File: 0143-reorder-list/0143-reorder-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
     void reorderList(ListNode* head) {
        if (!head || !head->next || !head->next->next) {
            return;
        }

        ListNode* slow = head;
        ListNode* fast = head;

        while (fast && fast->next) {
            fast = fast->next->next;
            slow = slow->next;
        }

        stack<ListNode*> st;
        ListNode* curr = slow->next;
        slow->next = nullptr;

        while (curr) {
            st.push(curr);
            curr = curr->next;
        }

        curr = head;
        ListNode* nextNode = curr->next;

        while (!st.empty()) {
            ListNode* top = st.top();
            st.pop();
            curr->next = top;
            top->next = nextNode;
            curr = nextNode;

            if (nextNode == nullptr) {
                break;
            }

            nextNode = nextNode->next;
        }
     }
};

File: 0143-reorder-list/0143-reorder-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverse(ListNode head){
        if(head == null || head.next == null)   return head;
        
        ListNode newHead = reverse(head.next);
        ListNode tail = head.next;
        
        head.next = null;
        tail.next = head;
        
        return newHead;
    }
    public void reorderList(ListNode head) {
        int n = 0;
        ListNode prev = head;

        while(prev != null){
            prev = prev.next;
            ++n;
        }

        prev = null;
        ListNode slow = head;
        ListNode fast = head;
        ListNode head2 = null;

        while(fast != null && fast.next != null){
            prev = slow;
            slow = slow.next;
            fast = fast.next.next;
        }

        if(n % 2 == 0){
            head2 = slow;
            prev.next = null;
        }else{
            head2 = slow.next;
            slow.next = null;
        }

        head2 = reverse(head2);

        ListNode res = head, tail = head;
        head = head.next;

        boolean turn = false;

        while(head != null || head2 != null){ // Corrected condition
            if(turn){
                tail.next = head;
                tail = head;
                head = head.next;
            }else{
                tail.next = head2;
                tail = head2;
                head2 = head2.next;
            }

            turn = !turn;
        }

        head = res;

        return;
    }

}